Code/Tools/TestImpactFramework/Runtime/Common/Code/Include/Static/BuildTarget/Common/TestImpactBuildGraph.h

```diff
@@ -160,6 +205,101 @@ namespace TestImpact
 
         return *vertex;
     }
+
+    template<typename ProductionTarget, typename TestTarget>
+    void BuildGraph<ProductionTarget, TestTarget>::WalkTargetBuildGraphSet(
+        TargetBuildGraph<ProductionTarget, TestTarget> BuildGraphVertex<ProductionTarget, TestTarget>::*graph,
+        TargetBuildGraphSet<ProductionTarget, TestTarget> TargetBuildGraph<ProductionTarget, TestTarget>::*set,
+        const BuildTarget<ProductionTarget, TestTarget>& buildTarget,
+        const BuildGraphVertexVisitor<ProductionTarget, TestTarget>& visitor) const
+    {
+        AZStd::queue<const BuildGraphVertex<ProductionTarget, TestTarget>*> vertexQueue;
+        AZStd::unordered_set<const BuildGraphVertex<ProductionTarget, TestTarget>*> visitedVertices;
+        AZStd::unordered_map<const BuildGraphVertex<ProductionTarget, TestTarget>*, size_t> vertexDistances;
+
+        // Skip visiting the root vertex and start visiting its children instead
+        const auto& parentVertex = GetVertexOrThrow(buildTarget);
+        visitedVertices.insert(&parentVertex);
+        for (const auto* child : parentVertex.*graph.*set)
+        {
+            vertexQueue.push(child);
+            vertexDistances[child] = 1;
+        }
+
+        while (!vertexQueue.empty())
+        {
+            const auto* vertex = vertexQueue.front();
+            vertexQueue.pop();
+
+            if (const auto result = visitor(*vertex, vertexDistances[vertex]);
+                result == BuildGraphVertexVisitResult::AbortGraphTraversal)
+            {
+                return;
```

Comment:
I'm not sure if there's a way we could do this consistently and unambiguously, i.e. if we abort a branch, but finish the traversal, what would we return? I see your point re: testing as it would be useful info nothing that couldn't be implemented using the callback itself (where the context of the traversal could be recorded appropriately for test validation etc.).